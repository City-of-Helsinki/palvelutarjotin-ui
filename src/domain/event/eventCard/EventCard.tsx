/* eslint-disable jsx-a11y/anchor-is-valid */
import { Button, IconLocation } from 'hds-react';
import times from 'lodash/times';
import { TFunction, useTranslation } from 'next-i18next';
import Link from 'next/link';
import React from 'react';

import KeywordsList from '../../../common/components/keyword/KeywordsList';
import SkeletonLoader from '../../../common/components/skeletonLoader/SkeletonLoader';
import {
  EventsFieldsFragment,
  useOccurrencesQuery,
} from '../../../generated/graphql';
import useLocale from '../../../hooks/useLocale';
import IconClock from '../../../icons/IconClock';
import { Language } from '../../../types';
import getLocalisedString from '../../../utils/getLocalisedString';
import { formatDateRange } from '../../../utils/time/format';
import { isMultidayOccurrence } from '../../occurrence/utils';
import PlaceText from '../../place/placeText/PlaceText';
import { getEventPlaceholderImage, getEventStartTimeStr } from '../utils';
import styles from './eventCard.module.scss';

interface Props {
  event: EventsFieldsFragment;
  link: string;
}

const EventCard: React.FC<Props> = ({ event, link }) => {
  const locale = useLocale();
  const id = event.id;
  const name = getLocalisedString(event.name, locale);
  const shortDescription = getLocalisedString(
    event.shortDescription || {},
    locale
  );
  const image = event.images[0]?.url;
  const keywords = [...event.keywords, ...event.audience];

  return (
    <Link href={link}>
      <a
        className={styles.eventCard}
        // TODO: should we use this? maybe not, the screen reader might not read everything
        // aria-label={t('event:eventCard.ariaLabelOpenEvent', {
        //   eventName: name,
        // })}
      >
        <div
          className={styles.imageWrapper}
          style={{
            backgroundImage: `url(${
              image || getEventPlaceholderImage(id || '')
            })`,
          }}
        ></div>
        <div className={styles.contentWrapper}>
          <div className={styles.titleWrapper}>
            <div className={styles.title}>{name}</div>
            <div className={styles.description}>{shortDescription}</div>
          </div>
          <div className={styles.occurrenceInfoWrapper}>
            <EventTime event={event} />
            <EventPlaceInfo event={event} />
          </div>
          <KeywordsList keywords={keywords} />
        </div>
      </a>
    </Link>
  );
};

const EventPlaceInfo: React.FC<{
  event: EventsFieldsFragment;
}> = ({ event }) => {
  return (
    <div className={styles.textWithIcon}>
      <IconLocation />
      <PlaceText placeId={event.location?.id || ''} />
    </div>
  );
};

const EventTime: React.FC<{
  event: EventsFieldsFragment;
}> = ({ event }) => {
  const { t } = useTranslation();
  const [showOccurrences, setShowOccurrences] = React.useState(false);

  const toggleShowOccurrences = (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    e.preventDefault(); // Prevent event card opening
    setShowOccurrences((showOccurrences) => !showOccurrences); // Toggle occurrences view
  };
  const nextOccurrence = event.pEvent.nextOccurrence?.edges[0]?.node;
  const nextOccurrenceTime = event.pEvent.nextOccurrenceDatetime;
  const lastOccurrenceTime =
    event.pEvent.lastOccurrenceDatetime ?? nextOccurrenceTime;
  const hasMultipleFutureOccurrences =
    nextOccurrenceTime && nextOccurrenceTime !== lastOccurrenceTime;

  const { data: occurrencesData, loading: loadingOccurrencesData } =
    useOccurrencesQuery({
      skip: !showOccurrences, // Fetch and render occurrences only when expanded
      variables: {
        cancelled: false,
        pEvent: event.pEvent.id,
        orderBy: 'startTime',
        upcoming: true,
      },
    });

  const occurrenceTimes = loadingOccurrencesData
    ? times(3, (key) => (
        <li key={`skeleton-${key}`}>
          <SkeletonLoader />
        </li>
      ))
    : occurrencesData?.occurrences?.edges
        ?.slice(1) // First one is already rendered
        .map((edge) =>
          edge?.node ? (
            <OccurrenceTime
              key={`occurrence-${edge?.node?.id}`}
              occurrence={edge?.node}
            />
          ) : null
        );

  return (
    <div className={styles.eventTimes}>
      <ul className={styles.occurrenceTimes}>
        {nextOccurrence ? (
          <OccurrenceTime occurrence={nextOccurrence} />
        ) : (
          t('event:eventCard.noUpcomingOccurrences')
        )}
        {showOccurrences && occurrenceTimes}
      </ul>
      {hasMultipleFutureOccurrences && (
        <Button
          variant="secondary"
          className={styles.multipleOccurrenceButton}
          onClick={(e) => toggleShowOccurrences(e)}
        >
          {showOccurrences
            ? t('event:eventCard.buttonMultipleOccurrences.opened')
            : t('event:eventCard.buttonMultipleOccurrences.closed')}
        </Button>
      )}
    </div>
  );
};

const OccurrenceTime: React.FC<{
  occurrence: { startTime: string; endTime: string };
}> = ({ occurrence }) => {
  const { t } = useTranslation();
  const locale = useLocale();
  return (
    <li className={styles.textWithIcon}>
      <IconClock />
      {getOccurrenceDateStr(occurrence, locale, t)}
    </li>
  );
};

export const getOccurrenceDateStr = (
  occurrence: { startTime: string; endTime: string },
  locale: Language,
  t: TFunction
): string | null => {
  if (isMultidayOccurrence(occurrence)) {
    return formatDateRange(
      new Date(occurrence.startTime),
      new Date(occurrence.endTime)
    );
  }

  return getEventStartTimeStr(new Date(occurrence.startTime), locale, t);
};

export default EventCard;
