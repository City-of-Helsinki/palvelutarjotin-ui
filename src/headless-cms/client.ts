/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { ApolloClient, InMemoryCache } from '@apollo/client';

const apolloClient = new ApolloClient({
  uri: process.env.NEXT_PUBLIC_CMS_BASE_URL,
  cache: new InMemoryCache(),
});
export default apolloClient;
