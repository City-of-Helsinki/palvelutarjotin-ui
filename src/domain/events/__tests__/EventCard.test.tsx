import { MockedResponse } from '@apollo/client/testing';
import { advanceTo } from 'jest-date-mock';
import React from 'react';

import * as graphql from '../../../generated/graphql';
import { emptyKeywordSetRequestMocks } from '../../../tests/apollo-mocks/keywordSetMocks';
import {
  fakeEvent,
  fakeKeyword,
  fakeLocalizedObject,
  fakeOccurrences,
  fakePEvent,
} from '../../../utils/mockDataUtils';
import {
  act,
  render,
  screen,
  userEvent,
  waitFor,
} from '../../../utils/testUtils';
import EventCard from '../../event/eventCard/EventCard';

const getPlaceRequestMock = (id: string) => ({
  request: {
    query: graphql.PlaceDocument,
    variables: {
      id,
    },
  },
  result: {
    data: null,
  },
});

describe('EventCard', () => {
  describe('List of occurrences', () => {
    it('renders a button to view multiple occurrences when event has them', async () => {
      const occurrences = fakeOccurrences(3, [
        { startTime: new Date(2020, 8, 20, 10, 30).toISOString() },
        { startTime: new Date(2020, 9, 21, 10, 20).toISOString() },
        { startTime: new Date(2020, 10, 22, 12, 40).toISOString() },
      ]);
      const event = fakeEvent({
        keywords: [
          fakeKeyword({
            name: fakeLocalizedObject('Avainsana'),
            id: 'kultus:1',
          }),
        ],
        pEvent: fakePEvent({
          nextOccurrenceDatetime: occurrences.edges[0]?.node?.startTime,
          lastOccurrenceDatetime: occurrences.edges[2]?.node?.startTime,
          nextOccurrence: fakeOccurrences(1, [
            {
              startTime: new Date(2020, 8, 20, 10, 30).toISOString(),
            },
          ]),
        }),
      });
      const apolloMocks: MockedResponse[] = [
        {
          request: {
            query: graphql.OccurrencesDocument,
            variables: {
              cancelled: false,
              pEvent: event.pEvent.id,
              orderBy: 'startTime',
            },
          },
          result: {
            data: {
              occurrences: occurrences,
            },
          },
        },
        {
          request: {
            query: graphql.OccurrencesDocument,
            variables: {
              cancelled: false,
              pEvent: event.pEvent.id,
              orderBy: 'startTime',
              upcoming: true,
            },
          },
          result: {
            data: {
              occurrences: occurrences,
            },
          },
        },
        ...emptyKeywordSetRequestMocks,
        getPlaceRequestMock(event.location?.id ?? ''),
      ];

      render(<EventCard event={event} link={'#'} />, { mocks: apolloMocks });
      await waitFor(() => {
        expect(
          screen.getByText(/sunnuntaina 20\.9\. klo 10:30/)
        ).toBeInTheDocument();
      });
      expect(screen.queryByText(/21.9 klo 10:20/)).not.toBeInTheDocument();
      expect(screen.queryByText(/22.10 klo 12:40/)).not.toBeInTheDocument();

      await act(async () => {
        await userEvent.click(
          await screen.findByRole('button', {
            name: /\+ muita tapahtuma-aikoja/i,
          })
        );
      });

      await waitFor(() => {
        expect(screen.queryByTestId('loading-spinner')).not.toBeInTheDocument();
      });

      await screen.findByRole('button', {
        name: /piilota muut ajat/i,
      });

      await waitFor(() => {
        expect(screen.getByText(/klo 10:20/)).toBeInTheDocument();
        expect(screen.getByText(/klo 12:40/)).toBeInTheDocument();
      });
      expect(
        screen.getByText(/sunnuntaina 20\.9\. klo 10:30/)
      ).toBeInTheDocument();
    });

    it('does not render a button to view multiple occurrences when only 1 occurrence presents', () => {
      const occurrences = fakeOccurrences(1);
      const event = fakeEvent({
        pEvent: fakePEvent({ occurrences }),
      });

      render(
        <EventCard
          event={event}
          link={`fi/event/${event.pEvent.linkedEventId}`}
        />,
        {
          mocks: [
            ...emptyKeywordSetRequestMocks,
            getPlaceRequestMock(event.location?.id ?? ''),
          ],
        }
      );

      expect(
        screen.queryByRole('button', {
          name: /\+ muita tapahtuma-aikoja/i,
        })
      ).not.toBeInTheDocument();
    });

    it('renders multiday occurrence time correctly', async () => {
      const nextOccurrence = {
        startTime: new Date(2020, 8, 20, 10, 30).toISOString(),
        endTime: new Date(2020, 8, 22, 10, 30).toISOString(),
      };
      const occurrences = fakeOccurrences(3, [
        nextOccurrence,
        {
          startTime: new Date(2020, 9, 21, 10, 20).toISOString(),
          endTime: new Date(2020, 9, 28, 10, 20).toISOString(),
        },
        {
          startTime: new Date(2020, 10, 22, 12, 40).toISOString(),
          endTime: new Date(2020, 10, 25, 12, 40).toISOString(),
        },
      ]);
      const event = fakeEvent({
        name: fakeLocalizedObject('Nimi'),
        shortDescription: fakeLocalizedObject('Kuvaus'),
        keywords: [
          fakeKeyword({
            name: fakeLocalizedObject('Avainsana'),
            id: 'kultus: 1',
          }),
        ],
        pEvent: fakePEvent({
          nextOccurrenceDatetime: occurrences.edges[0]?.node?.startTime,
          lastOccurrenceDatetime: occurrences.edges[2]?.node?.startTime,
          nextOccurrence: fakeOccurrences(1, [nextOccurrence]),
        }),
      });
      const apolloMocks: MockedResponse[] = [
        {
          request: {
            query: graphql.OccurrencesDocument,
            variables: {
              cancelled: false,
              pEvent: event.pEvent.id,
              orderBy: 'startTime',
            },
          },
          result: {
            data: {
              occurrences: occurrences,
            },
          },
        },
        {
          request: {
            query: graphql.OccurrencesDocument,
            variables: {
              cancelled: false,
              pEvent: event.pEvent.id,
              orderBy: 'startTime',
              upcoming: true,
            },
          },
          result: {
            data: {
              occurrences: occurrences,
            },
          },
        },
        ...emptyKeywordSetRequestMocks,
        getPlaceRequestMock(event.location?.id ?? ''),
      ];

      const { container } = render(<EventCard event={event} link={'#'} />, {
        mocks: apolloMocks,
      });

      await waitFor(() => {
        expect(screen.getByText('20.9.2020 – 22.9.2020')).toBeInTheDocument();
      });

      await act(async () => {
        await userEvent.click(
          await screen.findByRole('button', {
            name: /\+ muita tapahtuma-aikoja/i,
          })
        );
      });

      await waitFor(() => {
        expect(screen.queryByTestId('loading-spinner')).not.toBeInTheDocument();
      });

      await screen.findByRole('button', {
        name: /piilota muut ajat/i,
      });

      screen.getByText('21.10.2020 – 28.10.2020');
      screen.getByText('22.11.2020 – 25.11.2020');

      expect(container).toMatchSnapshot();
    });
  });
  describe('dates', () => {
    it('renders date as tomorrow and ignores enrolment days', async () => {
      advanceTo(new Date(2020, 8, 19));
      const occurrences = fakeOccurrences(3, [
        { startTime: new Date(2020, 8, 20, 10, 30).toISOString() },
      ]);
      const event = fakeEvent({
        pEvent: fakePEvent({
          nextOccurrence: occurrences,
          enrolmentEndDays: 10,
        }),
      });
      render(<EventCard event={event} link={'#'} />, {
        mocks: [
          ...emptyKeywordSetRequestMocks,
          getPlaceRequestMock(event.location?.id ?? ''),
        ],
      });
      await waitFor(() => {
        expect(screen.getByText(/huomenna klo 10:30/i)).toBeInTheDocument();
      });
    });
  });
});
