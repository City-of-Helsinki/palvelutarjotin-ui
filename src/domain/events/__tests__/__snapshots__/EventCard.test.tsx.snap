// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventCard List of occurrences renders multiday occurrence time correctly 1`] = `
<div>
  <a
    aria-label="Siirry tapahtumaan: Nimi"
    class="eventCard"
    href="/#"
  >
    <div
      class="imageWrapper"
      id="event-card:palvelutarjotin:a0775a20-468a-437b-91ee-f7a9cabea321"
      style="background-image: url(https://api.hel.fi/linkedevents-test/media/images/test.png);"
    />
    <div
      class="contentWrapper"
    >
      <div
        class="titleWrapper"
      >
        <div
          class="title"
        >
          Nimi
        </div>
        <div
          class="description"
        >
          Kuvaus
        </div>
      </div>
      <div
        class="occurrenceInfoWrapper"
      >
        <div
          class="eventTimes"
        >
          <ul
            class="occurrenceTimes"
          >
            <li
              class="textWithIcon"
            >
              <svg
                class=""
                height="100%"
                viewBox="0 0 24 24"
                width="100%"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill-rule="evenodd"
                  stroke-width="1"
                >
                  <path
                    d="M11.80416,-2.30926389e-13 C18.32352,-2.30926389e-13 23.60928,5.2848 23.60928,11.80416 C23.60928,18.32352 18.32352,23.60928 11.80416,23.60928 C5.2848,23.60928 -3.37507799e-14,18.32352 -3.37507799e-14,11.80416 C-3.37507799e-14,5.2848 5.2848,-2.30926389e-13 11.80416,-2.30926389e-13 Z M11.80416,2.5296 C6.6912,2.5296 2.53056,6.69024 2.53056,11.80416 C2.53056,16.91904 6.6912,21.07968 11.80416,21.07968 C16.91808,21.07968 21.07968,16.91904 21.07968,11.80416 C21.07968,6.69024 16.91808,2.5296 11.80416,2.5296 Z M13.068768,5.842464 L13.068768,11.280864 L16.665888,14.877024 L14.878368,16.665504 L10.539168,12.328224 L10.539168,5.842464 L13.068768,5.842464 Z"
                  />
                </g>
              </svg>
              20.9.2020 – 22.9.2020
            </li>
            <li
              class="textWithIcon"
            >
              <svg
                class=""
                height="100%"
                viewBox="0 0 24 24"
                width="100%"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill-rule="evenodd"
                  stroke-width="1"
                >
                  <path
                    d="M11.80416,-2.30926389e-13 C18.32352,-2.30926389e-13 23.60928,5.2848 23.60928,11.80416 C23.60928,18.32352 18.32352,23.60928 11.80416,23.60928 C5.2848,23.60928 -3.37507799e-14,18.32352 -3.37507799e-14,11.80416 C-3.37507799e-14,5.2848 5.2848,-2.30926389e-13 11.80416,-2.30926389e-13 Z M11.80416,2.5296 C6.6912,2.5296 2.53056,6.69024 2.53056,11.80416 C2.53056,16.91904 6.6912,21.07968 11.80416,21.07968 C16.91808,21.07968 21.07968,16.91904 21.07968,11.80416 C21.07968,6.69024 16.91808,2.5296 11.80416,2.5296 Z M13.068768,5.842464 L13.068768,11.280864 L16.665888,14.877024 L14.878368,16.665504 L10.539168,12.328224 L10.539168,5.842464 L13.068768,5.842464 Z"
                  />
                </g>
              </svg>
              21.10.2020 – 28.10.2020
            </li>
            <li
              class="textWithIcon"
            >
              <svg
                class=""
                height="100%"
                viewBox="0 0 24 24"
                width="100%"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill-rule="evenodd"
                  stroke-width="1"
                >
                  <path
                    d="M11.80416,-2.30926389e-13 C18.32352,-2.30926389e-13 23.60928,5.2848 23.60928,11.80416 C23.60928,18.32352 18.32352,23.60928 11.80416,23.60928 C5.2848,23.60928 -3.37507799e-14,18.32352 -3.37507799e-14,11.80416 C-3.37507799e-14,5.2848 5.2848,-2.30926389e-13 11.80416,-2.30926389e-13 Z M11.80416,2.5296 C6.6912,2.5296 2.53056,6.69024 2.53056,11.80416 C2.53056,16.91904 6.6912,21.07968 11.80416,21.07968 C16.91808,21.07968 21.07968,16.91904 21.07968,11.80416 C21.07968,6.69024 16.91808,2.5296 11.80416,2.5296 Z M13.068768,5.842464 L13.068768,11.280864 L16.665888,14.877024 L14.878368,16.665504 L10.539168,12.328224 L10.539168,5.842464 L13.068768,5.842464 Z"
                  />
                </g>
              </svg>
              22.11.2020 – 25.11.2020
            </li>
          </ul>
          <button
            class="Button-module_button__1msFE button_hds-button__2A0je Button-module_secondary__1nABp button_hds-button--secondary__1NOWS multipleOccurrenceButton"
            type="button"
          >
            <span
              class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
            >
              piilota muut ajat
            </span>
          </button>
        </div>
        <div
          class="textWithIcon"
        >
          <svg
            aria-hidden="true"
            aria-label="location"
            class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M11.9669 1.5C14.0265 1.5 16.0869 2.27765 17.6579 3.83367C20.8001 6.945 20.5878 11.7938 17.6579 15.1017L17.0345 15.8107C14.4235 18.8021 12.9688 20.7706 11.9669 22.7477C10.8937 20.6185 9.28527 18.4993 6.27797 15.1017C3.34814 11.7938 3.13508 6.945 6.27797 3.83367C7.84906 2.27765 9.90869 1.5 11.9669 1.5ZM11.999 3.5C10.3743 3.5 8.84685 4.12598 7.69956 5.26179C6.56645 6.38341 5.97897 7.81769 6.00058 9.41066C6.02151 10.984 6.64884 12.5453 7.76709 13.8087L8.42773 14.5606C10.0167 16.3842 11.1442 17.8 12.0004 19.0644C12.9748 17.6274 14.2916 16.0006 16.233 13.8087C17.3512 12.5453 17.9786 10.984 17.9995 9.41066C18.0198 7.81769 17.4323 6.38341 16.3005 5.26179C15.1519 4.12598 13.6251 3.5 11.999 3.5ZM12.0004 6C13.9338 6 15.5 7.56702 15.5 9.50039C15.5 11.433 13.9338 13 12.0004 13C10.067 13 8.5 11.433 8.5 9.50039C8.5 7.56702 10.067 6 12.0004 6ZM12.0003 8C11.1725 8 10.5 8.67319 10.5 9.50034C10.5 10.3268 11.1725 11 12.0003 11C12.8275 11 13.5 10.3268 13.5 9.50034C13.5 8.67319 12.8275 8 12.0003 8Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
      <ul
        class="keywordsList"
      >
        <li>
          <button
            aria-label="Hae tapahtumia avainsanalla Avainsana"
            class="keyword"
          >
            Avainsana
          </button>
        </li>
      </ul>
    </div>
  </a>
</div>
`;
